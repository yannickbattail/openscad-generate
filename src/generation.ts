import * as path from "node:path";
import * as fs from "node:fs";
import chalk from "chalk";
import pLimit, { LimitFunction } from "p-limit";

import { createFctExecCommand } from "./util/execBash.js";
import { GenerateOptions } from "./types.js";
import { OpenScad, OpenScadOutputWithSummary, ParameterSet } from "openscad-cli-wrapper";
import { genParamSetInFormat } from "./generateFormat.js";
import { GenerateMosaic } from "./util/MosaicGeneration.js";
import JSON5 from "json5";

export async function generate(genOptions: GenerateOptions) {
  const executor = createFctExecCommand(!genOptions.openScadOptions.debug, !!genOptions.openScadOptions.debug);
  const limiter: LimitFunction = pLimit(genOptions.parallelJobs);
  console.log(chalk.green(`Generating model for file: ${genOptions.fileName} in formats: ${genOptions.outFormats}`));

  if (!fs.existsSync(genOptions.outputDir)) {
    fs.mkdirSync(genOptions.outputDir);
  }

  const openscad = new OpenScad(genOptions.fileName, genOptions.outputDir, executor);
  try {
    const { parameterSetFileName, paramSetToGenerate } = fetchParameterSets(genOptions);
    const tasks: Promise<OpenScadOutputWithSummary | null>[] = paramSetToGenerate
      .map((paramSet) =>
        genOptions.outFormats.map((format) =>
          limiter(async () =>
            genParamSetInFormat(
              format,
              openscad,
              {
                parameterFile: parameterSetFileName,
                parameterName: paramSet[0],
              },
              genOptions,
              executor,
            ),
          ),
        ),
      )
      .flat();
    const result = await Promise.allSettled(tasks);
    for (const fail of result.filter((r) => r.status === "rejected").map((r) => r.reason)) {
      console.error(chalk.red(`💥 Error generating parameter set`, fail));
    }
    if (genOptions.generateMosaic) {
      const pngFiles = getPngResult(result.filter((r) => r.status === "fulfilled").map((r) => r.value));
      if (pngFiles) {
        await GenerateMosaic(pngFiles, genOptions, executor);
      } else {
        console.error("⚠️ No PNG files for generating mosaic. (you need to generate PNG images)");
      }
    }
  } catch (error) {
    console.error(`💥 Error generating parameter set`, error);
  }
}

function fetchParameterSets(genOptions: GenerateOptions) {
  const fileNameWithoutExtension = path.basename(genOptions.fileName, path.extname(genOptions.fileName));
  const parameterSetFileName = fileNameWithoutExtension + ".json";
  const fileContent = fs.readFileSync(fileNameWithoutExtension + ".json", "utf-8");
  const parameterSet: ParameterSet = JSON5.parse(fileContent) satisfies ParameterSet;
  const paramSetToGenerate = Object.entries(parameterSet.parameterSets).filter(
    (paramSet) => !genOptions.onlyParameterSet || genOptions.onlyParameterSet === paramSet[0],
  );
  return { parameterSetFileName, paramSetToGenerate };
}

function getPngResult(result: (OpenScadOutputWithSummary | null)[]): string[] {
  return result
    .filter((r) => r !== null)
    .filter((r) => r.file.endsWith(".png"))
    .map((r) => r.file);
}
