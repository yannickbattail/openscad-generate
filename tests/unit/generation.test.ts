import { describe, expect, it } from "vitest";
import { getDefaultOpenscadOptions } from "../../src/generation.js";

describe("generation", () => {
  it("getOpenscadOptions", () => {
    const openScad = getDefaultOpenscadOptions();
    expect(openScad).toMatchSnapshot();
  });
});
