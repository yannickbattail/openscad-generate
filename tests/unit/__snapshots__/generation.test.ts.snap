// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generation > getOpenscadOptions 1`] = `
{
  "fileName": "",
  "generateMosaic": false,
  "mosaicOptions": {
    "debug": false,
    "geometry": {
      "border": 2,
      "height": 1024,
      "width": 1024,
    },
    "outputPath": "",
    "scadFileName": "",
    "tiles": {
      "height": 2,
      "width": 2,
    },
  },
  "onlyParameterSet": "",
  "openScadOptions": {
    "animOptions": {
      "animDelay": 50,
      "animate": 50,
      "animate_sharding": null,
      "autocenter": false,
      "camera": null,
      "colorscheme": "DeepOcean",
      "csglimit": null,
      "imgsize": {
        "height": 512,
        "width": 515,
      },
      "preview": null,
      "projection": null,
      "render": null,
      "view": null,
      "viewall": false,
    },
    "backend": "Manifold",
    "check_parameter_ranges": false,
    "check_parameters": false,
    "debug": null,
    "experimentalFeatures": {
      "import_function": true,
      "input_driver_dbus": false,
      "lazy_union": true,
      "predictible_output": true,
      "python_engine": false,
      "roof": true,
      "textmetrics": true,
      "vertex_object_renderers_indexing": true,
    },
    "hardwarnings": false,
    "imageOptions": {
      "autocenter": false,
      "camera": null,
      "colorscheme": "DeepOcean",
      "csglimit": null,
      "imgsize": {
        "height": 1024,
        "width": 1024,
      },
      "preview": null,
      "projection": null,
      "render": null,
      "view": null,
      "viewall": false,
    },
    "openScadExecutable": "openscad",
    "option3mf": {
      "add_meta_data": "true",
      "color": "",
      "color_mode": "model",
      "decimal_precision": "6",
      "material_type": "basematerial",
      "meta_data_copyright": "",
      "meta_data_description": "",
      "meta_data_designer": "",
      "meta_data_license_terms": "",
      "meta_data_rating": "",
      "meta_data_title": "",
      "unit": "millimeter",
    },
    "python_module": null,
    "quiet": false,
    "trust_python": false,
  },
  "outFormats": [
    "png",
    "webp",
    "3mf",
  ],
  "parallelJobs": 1,
}
`;
